### PR TITLE
:warning: Disable the `rest_client_request_latency_seconds` metric by default

### LINKED ISSUE #1423: Disable rest client latency metric
We vendor in `controller-runtime` and use it to expose its controller specific metrics in addition to our own.
Recently we observed that [rest_client_request_latency_seconds](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/metrics/client_go_adapter.go#L55-L60) was eating up a significant portion of Prometheus storage, as can be seen from a snapshot of the highest cardinality metrics in our codebase:
```
42765 kube_secret_type
42799 kube_secret_metadata_resource_version
461312 rest_client_request_latency_seconds_count
461312 rest_client_request_latency_seconds_sum
5074432 rest_client_request_latency_seconds_bucket
```

I understand this metric was deprecated in kube 1.14 and turned off in [1.17](https://github.com/kubernetes/kubernetes/pull/83836).
Is it possible to either disable this metric or to allow us to unregister it?

### PR DESCRIPTION
**Pull request by 2uasimojo**
This metric has been deprecated since kube 1.14 and [turned off in kube 1.17](https://github.com/kubernetes/kubernetes/pull/83836). Stop registering it by default, but export the appropriate variables so
consumers can register it explicitly if they wish.

fixes #1423

### CODE REVIEW

File: pkg/metrics/client_go_adapter.go [hunk 1]
```diff
@@ -52,7 +52,21 @@ const (
 
 var (
 	// client metrics.
-	requestLatency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+
+	// RequestLatency reports the request latency in seconds per verb/URL.
```
Comment (alvaroaleman) [inline pkg/metrics/client_go_adapter.go:56]: Please document that this is deprecated, going to get removed and what the issue with it is.

**Review by logicalhan — COMMENTED** [review]
what is this? this isn't really making it optional.. you should have a command line flag which allows you to enable this metric and plumb the value of that flag to the metric registration function

**Comment by alvaroaleman** [comment]
> what is this? this isn't really making it optional.. you should have a command line flag which allows you to enable this metric and plumb the value of that flag to the metric registration function

No. We are a library. We are not going to register a flag. How and if controller authors expose options in their controllers to end-users is up to them.

**Comment by logicalhan** [comment]
component-base is also a library, it includes flags which can be used by consuming binaries. these include the ones for metrics.

**Comment by 2uasimojo** [comment]
...but I'm with you on the plumbing part. If there was a way we could set a package variable such that it's present by the time the import `init()`s, we could switch it on/off from client code. Is that even possible in go?

**Comment by logicalhan** [comment]
> ...but I'm with you on the plumbing part. If there was a way we could set a package variable such that it's present by the time the import `init()`s, we could switch it on/off from client code. Is that even possible in go?

yes it is possible, we literally do this in upstream kubernetes.

**Comment by logicalhan** [comment]
obviously this means you can't use the init for these though.

**Comment by 2uasimojo** [comment]
Okay; I was looking for a way to enable opt-out so the change could be non-breaking. But if

> you can't use the init for these

then I'm not sure how that would work.

**Comment by logicalhan** [comment]
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/metrics/metrics.go#L506-L546

...is one way.

**Comment by alvaroaleman** [comment]
Kubelet is a binary, we are a library. I am going to veto any change that registers a flag. The one case where we do this creates enough trouble already: https://github.com/kubernetes-sigs/controller-runtime/issues/1396

**Comment by logicalhan** [comment]
> Okay; I was looking for a way to enable opt-out so the change could be non-breaking. 

I think it's okay to make it opt-in, it's more friendly than just deleting it. And the approach of manually invoking Register is now an option.

**Comment by logicalhan** [comment]
> Kubelet is a binary, we are a library. I am going to veto any change that registers a flag. The one case where we do this creates enough trouble already: #1396

That's fine, I don't have any skin in the game here, I am just saying that it's not friendly to outright break people. What you guys want to do with your library and what your priorities are 🤷 to me.. I'm stating my opinion about deleting metrics and breaking people.. I've had it happen to me and it sucks.

**Comment by logicalhan** [comment]
Actually, as a library, it would make sense to make all your metrics opt-in...

**Comment by 2uasimojo** [comment]
So if we don't

> registers a flag

and we want to

> make it opt-in

then is the approach in this PR acceptable? Or would it be better to wrap the registration of this metric in its own `Register` func?

Alternatively, I would be happy to propose

> make all your metrics opt-in

**Comment by logicalhan** [comment]
Yeah I would probably make all the metrics public and auto-register them for a release and have a release note saying that in the next release, these will all be opt in. That way, people have some sort of warning, rather than just randomly have everything break. You have several options (as I see it):

- expose all metrics individually (this is kinda a pain because now you have to know all the ones you want to register)
- expose groups of metrics (this is slightly easier for consumers)
- expose a single Register which registers all metrics (possibly taking a var args of metrics not to enable).

**Comment by 2uasimojo** [comment]
I'm happy to do anything in this range, from [outright removal](https://github.com/kubernetes-sigs/controller-runtime/pull/1586) to immediate opt-in (this PR) to a multi-release no-change-deprecation => opt-in dance.

@alvaroaleman what is your wish?

**Comment by alvaroaleman** [comment]
This change modulo my comment https://github.com/kubernetes-sigs/controller-runtime/pull/1587/files#r667037223 is IMHO fine. I don't think we should require opt-in for every single metric but I am happy to discuss if, which and how we register metrics by default in a new issue.

File: pkg/metrics/client_go_adapter.go (hunk 1 above)
Comment (2uasimojo) [reply pkg/metrics/client_go_adapter.go:56]: Done.

**Comment by 2uasimojo** [comment]
> This change modulo my comment

✓

> I don't think we should require opt-in for every single metric but I am happy to discuss

See https://github.com/kubernetes-sigs/controller-runtime/pull/1589 for one approach: opt in gets you all by default, but you have the ability to opt *out* of individual metrics.

**Comment by alvaroaleman** [comment]
I am going to retitle this to make sure that its properly understood when it ends up in a release note:
/retitle :warning: Disable the `rest_client_request_latency_seconds` metric by default

**Review by alvaroaleman — APPROVED** [review]
thanks!

**Comment by 2uasimojo** [comment]
Okay, so what's the process for getting this into a tagged release?

**Comment by alvaroaleman** [comment]
@2uasimojo we deliver breaking changes with a new minor release and do a new minor release whenever there is a new kube minor release. So this will take some time.

**Pull request merged**